### ISSUE TITLE
Option for acknowledging terminating pods in rolling update

### ISSUE THREAD

**Issue by atiratree** [kubernetes/kubernetes#107920]
### What would you like to be added?

It could make sense to wait for a pod to be terminated before scheduling a new one for the issues mentioned bellow. Even though some issues can be partially mitigated by proper setup of maxUnavailable and maxSurge, it is not applicable for all of them.

I would like to propose a new opt-in behaviour that would solve this. Deployment controller would include Terminating pods in the computation of current running replicas when deciding if the new RS should scale up (or old in case of proportional scaling).

This could be configured for example in `.spec.strategy.rollingUpdate.scalingPolicy` with possible values
1. `IgnoreTerminatingPods` - default and current behaviour
2. `WaitForTerminatingPods`

The disadvantage of this feature is a slower rollout in resource unrestricted environments. So, using this feature would be advised only for similar use cases to the mentioned ones above.

### Why is this needed?

In some cases people are surprised that their deployment can momentarily have more pods during a rollout than described `( replicas - maxUnavailable < availableReplicas < replicas + maxSurge)`. The culprit are Terminating pods that can run in addition to the Running + Starting pods.
Even though Terminating pods are not considered part of a deployment this can cause problems with resource usage and scheduling:

1. Unnecessary autoscaling of nodes in tight environments and driving up cloud costs. This can hurt especially if
    - you rollout multiple deployments at the same time.
    - you have generous termination periods and your pods take a long time to shutdown (example here https://github.com/kubernetes/kubernetes/issues/95498#issuecomment-814048997)
 
    relevant issues:
    - https://github.com/kubernetes/kubernetes/issues/95498
    - https://github.com/kubernetes/kubernetes/issues/99513
    - https://github.com/kubernetes/kubernetes/issues/41596
    - https://github.com/kubernetes/kubernetes/issues/97227
 
2. A problem also arises in contentious environments where pods are fighting for resources. This can bring up exponential backoff for not yet started pods into big numbers and unnecessarily delay start of such pods until they pop from the queue when there are computing resources to run them. This can slow down the deployment considerably.

    relevant issue: https://github.com/kubernetes/kubernetes/issues/98656

    In this issue the resources were limited by a quota, but this can be due to other reasons as well. In our use case we noticed, this can occur also in high availability scenarios where pods are expected to run only on certain nodes and pod anti-affinity forbids to run two pods at the same node.

**Comment by atiratree** [comment]
This was discussed in last sig-apps meeting with a following decision (https://www.youtube.com/watch?v=8hs2jDfbXsw): 
1. let's wait for this issue to get more traction before pursuing it further
2. document the current behaviour

**Comment by gjkim42** [comment]
How is this different from statefulset?
I guess the pods from statefulset are guaranteed to be unique so they only start after termination.

**Comment by atiratree** [comment]
Yes, exactly, stateful set waits for the termination by default. You can override that by setting `podManagementPolicy` to parallel. Please see https://github.com/kubernetes/enhancements/pull/3940 KEP for additional discussions.

**Comment by h0jeZvgoxFepBQ2C** [comment]
Any updates if this gets merged?

**Comment by atiratree** [comment]
I will try to pick this up soon.

**Comment by atiratree** [comment]
we are planning to work on the feature and targeting 1.30: https://github.com/kubernetes/enhancements/pull/4357

**Comment by atiratree** [comment]
the feature missed the last release, targeting 1.31 now: https://github.com/kubernetes/enhancements/pull/4670

**Comment by atiratree** [comment]
I plan to work on this in 1.34
/triage accepted
